cli/command/stack/kubernetes/deploy.go

```diff
@@ -61,10 +60,110 @@ func RunDeploy(dockerCli *KubeCli, opts options.Deploy) error {
 	}
 
 	fmt.Fprintln(cmdOut, "Waiting for the stack to be stable and running...")
+	v1beta1Cli, err := dockerCli.stacksv1beta1()
+	if err != nil {
+		return err
+	}
 
-	<-watcher.Watch(stack.name, stack.getServices())
-
-	fmt.Fprintf(cmdOut, "Stack %s is stable and running\n\n", stack.name)
+	pods := composeClient.Pods()
+	watcher := &deployWatcher{
+		stacks: v1beta1Cli,
+		pods:   pods,
+	}
+	statusUpdates := make(chan serviceStatus)
+	displayDone := make(chan struct{})
+	go func() {
+		defer close(displayDone)
+		display := newStatusDisplay(dockerCli.Out())
+		for status := range statusUpdates {
+			display.OnStatus(status)
+		}
+	}()
 
+	err = watcher.Watch(stack.name, stack.getServices(), statusUpdates)
+	close(statusUpdates)
+	<-displayDone
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(cmdOut, "\nStack %s is stable and running\n\n", stack.name)
 	return nil
+
+}
+
+type statusDisplay interface {
+	OnStatus(serviceStatus)
+}
+type metaServiceState string
+
+const (
+	metaServiceStateReady   = metaServiceState("Ready")
+	metaServiceStatePending = metaServiceState("Pending")
+	metaServiceStateFailed  = metaServiceState("Failed")
+)
+
+func metaStateFromStatus(status serviceStatus) metaServiceState {
+	switch {
+	case status.podsReady > 0:
+		return metaServiceStateReady
+	case status.podsPending > 0:
+		return metaServiceStatePending
+	default:
+		return metaServiceStateFailed
+	}
+}
+
+type forwardOnlyStatusDisplay struct {
+	o      *command.OutStream
+	states map[string]metaServiceState
+}
+
+func (d *forwardOnlyStatusDisplay) OnStatus(status serviceStatus) {
+	state := metaStateFromStatus(status)
+	if d.states[status.name] != state {
+		d.states[status.name] = state
+		fmt.Fprintf(d.o, "%s: %s", status.name, state)
+		fmt.Fprintln(d.o)
+	}
+}
+
+type interactiveStatusDisplay struct {
+	o        *command.OutStream
+	statuses []serviceStatus
+}
+
+func (d *interactiveStatusDisplay) OnStatus(status serviceStatus) {
+	b := aec.EmptyBuilder
+	for ix := 0; ix < len(d.statuses); ix++ {
+		b = b.Up(1).EraseLine(aec.EraseModes.All)
```

Comment:
The original issue which lead (eventually) to `Gotty` use was #28111 (fixed in #28238, Gotty in followup #28304) solving a problem with `Up(0)` (and `Down(0)` etc) being undefined, but that isn't used here and this `aec` library looks to do the right thing but is just using plain ANSI codes rather than terminfo, so may theoretically not be as portable, although in practice by sticking to ANSI it's probably going to work most reasonable places.

`Gotty` is not maintained (my PRs made as part of the above still aren't merged) whereas `aec` has been more recently (and has no issue or prs open). Maybe it'd be worth moving the `jsonmessage` stuff over to that.

Reply:
Thanks for looking @ijc - let me open an issue in moby/moby to discuss replacing Gotty

Reply:
Fyi, I chose this lib because of its simplicity, and because it is already used in buildkit (which has IMO a very modern and good looking UX)